// @allurereport/plugin-testops/dist/plugin.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TestopsUploaderPlugin_instances, _TestopsUploaderPlugin_ci, _TestopsUploaderPlugin_client, _TestopsUploaderPlugin_launchName, _TestopsUploaderPlugin_launchTags, _TestopsUploaderPlugin_uploadedTestResultsIds, _TestopsUploaderPlugin_upload, _TestopsUploaderPlugin_startUpload, _TestopsUploaderPlugin_stopUpload;
import { detect } from "@allurereport/ci";
import { getWorstStatus } from "@allurereport/core-api";
import { convertToSummaryTestResult, } from "@allurereport/plugin-api";
import { TestOpsClient } from "./client.js";
import { resolvePluginOptions, unwrapStepsAttachments } from "./utils.js";
export class TestopsUploaderPlugin {
    constructor(options) {
        _TestopsUploaderPlugin_instances.add(this);
        this.options = options;
        _TestopsUploaderPlugin_ci.set(this, void 0);
        _TestopsUploaderPlugin_client.set(this, void 0);
        _TestopsUploaderPlugin_launchName.set(this, "");
        _TestopsUploaderPlugin_launchTags.set(this, []);
        _TestopsUploaderPlugin_uploadedTestResultsIds.set(this, []);
        const { accessToken, endpoint, projectId, launchName, launchTags } = resolvePluginOptions(options);
        __classPrivateFieldSet(this, _TestopsUploaderPlugin_ci, detect(), "f");
        if ([accessToken, endpoint, projectId].every(Boolean)) {
            __classPrivateFieldSet(this, _TestopsUploaderPlugin_client, new TestOpsClient({
                baseUrl: endpoint,
                accessToken,
                projectId,
            }), "f");
            __classPrivateFieldSet(this, _TestopsUploaderPlugin_launchName, launchName, "f");
            __classPrivateFieldSet(this, _TestopsUploaderPlugin_launchTags, launchTags, "f");
        }
        if (!accessToken) {
            console.warn("TestOps access token is missing. Please provide a valid access token in the plugin options.");
        }
        if (!endpoint) {
            console.warn("TestOps endpoint is missing. Please provide a valid endpoint in the plugin options.");
        }
        if (!projectId) {
            console.warn("TestOps project ID is missing. Please provide a valid project ID in the plugin options.");
        }
    }
    get ciMode() {
        return __classPrivateFieldGet(this, _TestopsUploaderPlugin_ci, "f") && __classPrivateFieldGet(this, _TestopsUploaderPlugin_ci, "f").type !== "local";
    }
    async start(_context, store) {
        if (!__classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f")) {
            return;
        }
        await __classPrivateFieldGet(this, _TestopsUploaderPlugin_instances, "m", _TestopsUploaderPlugin_startUpload).call(this);
        await __classPrivateFieldGet(this, _TestopsUploaderPlugin_instances, "m", _TestopsUploaderPlugin_upload).call(this, store, { issueNewToken: false });
        console.info(`TestOps launch has been created: ${__classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f").launchUrl}`);
    }
    async update(_context, store) {
        if (!__classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f")) {
            return;
        }
        await __classPrivateFieldGet(this, _TestopsUploaderPlugin_instances, "m", _TestopsUploaderPlugin_upload).call(this, store);
    }
    async done(_context, store) {
        if (!__classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f")) {
            return;
        }
        const allTrs = (await store.allTestResults()).filter((tr) => this.options.filter ? this.options.filter(tr) : true);
        const worstStatus = getWorstStatus(allTrs.map(({ status }) => status));
        await __classPrivateFieldGet(this, _TestopsUploaderPlugin_instances, "m", _TestopsUploaderPlugin_upload).call(this, store);
        await __classPrivateFieldGet(this, _TestopsUploaderPlugin_instances, "m", _TestopsUploaderPlugin_stopUpload).call(this, worstStatus || "unknown");
    }
    async info(context, store) {
        if (!__classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f")?.launchUrl) {
            return undefined;
        }
        const allTrs = (await store.allTestResults()).filter((tr) => this.options.filter ? this.options.filter(tr) : true);
        const newTrs = await store.allNewTestResults();
        const retryTrs = allTrs.filter((tr) => !!tr?.retries?.length);
        const flakyTrs = allTrs.filter((tr) => !!tr?.flaky);
        const duration = allTrs.reduce((acc, { duration: trDuration = 0 }) => acc + trDuration, 0);
        const worstStatus = getWorstStatus(allTrs.map(({ status }) => status));
        const createdAt = allTrs.reduce((acc, { stop }) => Math.max(acc, stop || 0), 0);
        return {
            name: __classPrivateFieldGet(this, _TestopsUploaderPlugin_launchName, "f"),
            remoteHref: __classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f").launchUrl,
            stats: await store.testsStatistic(this.options.filter),
            status: worstStatus ?? "passed",
            duration,
            createdAt,
            plugin: "Awesome",
            newTests: newTrs.map(convertToSummaryTestResult),
            flakyTests: flakyTrs.map(convertToSummaryTestResult),
            retryTests: retryTrs.map(convertToSummaryTestResult),
            meta: {
                reportId: context.reportUuid,
            },
        };
    }
}
_TestopsUploaderPlugin_ci = new WeakMap(), _TestopsUploaderPlugin_client = new WeakMap(), _TestopsUploaderPlugin_launchName = new WeakMap(), _TestopsUploaderPlugin_launchTags = new WeakMap(), _TestopsUploaderPlugin_uploadedTestResultsIds = new WeakMap(), _TestopsUploaderPlugin_instances = new WeakSet(), _TestopsUploaderPlugin_upload = async function _TestopsUploaderPlugin_upload(store, options) {
    const { issueNewToken = true } = options ?? {};
    const allTrs = await store.allTestResults();
    const trsToUpload = allTrs.filter((tr) => {
        const uploaded = __classPrivateFieldGet(this, _TestopsUploaderPlugin_uploadedTestResultsIds, "f").includes(tr.id);
        if (this.options.filter) {
            return this.options.filter(tr) && !uploaded;
        }
        return !uploaded;
    });
    if (trsToUpload.length === 0) {
        return;
    }
    const allTrsWithAttachments = trsToUpload.map((tr) => {
        return {
            ...tr,
            steps: unwrapStepsAttachments(tr.steps),
        };
    });
    if (issueNewToken) {
        await __classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f").issueOauthToken();
    }
    await __classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f").createSession();
    await __classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f").uploadTestResults({
        trs: allTrsWithAttachments,
        attachmentsResolver: async (tr) => {
            const attachments = await store.attachmentsByTrId(tr.id);
            return await Promise.all(attachments.map(async (attachment) => {
                const content = await store.attachmentContentById(attachment.id);
                return {
                    originalFileName: attachment.originalFileName,
                    contentType: attachment.contentType,
                    content: await content?.readContent(async (s) => s),
                };
            }));
        },
        fixturesResolver: async (tr) => {
            const fxts = await store.fixturesByTrId(tr.id);
            return fxts.map((fxt) => ({
                ...fxt,
                type: fxt.type.toUpperCase(),
                steps: unwrapStepsAttachments(fxt.steps),
            }));
        },
    });
    __classPrivateFieldGet(this, _TestopsUploaderPlugin_uploadedTestResultsIds, "f").push(...allTrsWithAttachments.map((tr) => tr.id));
}, _TestopsUploaderPlugin_startUpload = async function _TestopsUploaderPlugin_startUpload() {
    if (!__classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f")) {
        return;
    }
    await __classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f").issueOauthToken();
    await __classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f").createLaunch(__classPrivateFieldGet(this, _TestopsUploaderPlugin_launchName, "f"), __classPrivateFieldGet(this, _TestopsUploaderPlugin_launchTags, "f"));
    if (!this.ciMode) {
        return;
    }
    await __classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f").startUpload(__classPrivateFieldGet(this, _TestopsUploaderPlugin_ci, "f"));
}, _TestopsUploaderPlugin_stopUpload = async function _TestopsUploaderPlugin_stopUpload(status) {
    if (!this.ciMode || !__classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f")) {
        return;
    }
    await __classPrivateFieldGet(this, _TestopsUploaderPlugin_client, "f").stopUpload(__classPrivateFieldGet(this, _TestopsUploaderPlugin_ci, "f"), status);
};
